### DISCUSSION TITLE
[General question] What is the intended configuration when using rocks.nvim as a package manager

### DISCUSSION THREAD

**Discussion by NStefan002** [lumen-oss/rocks.nvim#139] [Q&A] [answered]
So, let's use just one plugin as an example. I just published [2048.nvim](https://github.com/NStefan002/2048.nvim) to luarocks. I installed it with `:Rocks install 2048`. And here comes the part that I'm interested in. By default rocks.nvim sources the `lua/` directory of the plugin, but unlike lazy.nvim it doesn't automatically call the `setup` function (as far as I can tell). So I guess that the approach here would be something like this: I have one file with plugin declarations (`rocks.toml`) and configurations files (e.g. `~/.config/nvim/lua/plugins/2048.lua`) where I call the `setup` function and do customizations if I want to. I know that this could be done in a million different ways, but I just wanted to understand the intended way of doing things and try to adapt to them so I can see the full picture. Let me know if I misunderstood something.

**Note:** I read the `rocks-config.nvim` docs, so I have a few more questions about that.
1. Does this:
```toml
[config]
plugins_dir = "plugins/"
```
do the same thing as I mentioned above (the `~/.config/nvim/lua/plugins/...` stuff)? The difference could be that with rocks-config I don't have to manually source the configurations, but without it, I would need to have `require('plugins')` in `~/.config/nvim/init.lua` and then in `lua/plugins/init.lua` lines such as `require('plugins.2048')` to source the file that contains my plugin config.

2. I understood that the `auto_setup` option is meant to be used when you don't want to pass any arguments to the plugin's `setup` function (in other words you are satisfied with the default config, or the plugin does not have any configuration options). So is this the equivalent to `config = true` or `opts = {}` in lazy.nvim? I'm sorry for creating parallels to `lazy` but that's currently what most people (including me) are familiar with.

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave: 

With rocks.nvim, there is no "intended way" of configuring plugins.
We follow the UNIX philosophy: "Do one thing and do it well".

So as a plugin manager, rocks.nvim leaves configuration up to the user, or an external module like rocks-config.nvim.
If you're using rocks-config.nvim, it woks as you describe, and you don't have to manually source the plugins.
The module will look at your `rocks.toml` and search for modules in your `config.plugins_dir`, based on the names of the respective plugins.
If you have a config in your `plugins_dir`, but the plugin isn't installed, rocks-config.nvim won't try to source it.

The `auto_setup` option exists as a workaround for the fact that many Lua plugins follow the bad practise of requiring the user to call a `setup` function, even if there's no configuration needed (see [my blog post on the topic](https://mrcjkb.dev/posts/2023-08-22-setup.html)).
`auto_setup` is disabled by default (because it can have unwanted side effects).
Afaik, lazy.nvim behaves like this by default, and allows users to disable it by setting `config = false` (or something like that - I'm not sure about the details, as I've never used lazy.nvim myself).

If you don't use rocks-config.nvim, there are many different ways you can configure your plugins.

- One option is to have a `$XDG_CONFIG_HOME/$NVIM_APPNAME/lua/plugins`[^1] module, and to `require` it in your `init.lua`.
- Another is to put files in the `$XDG_CONFIG_HOME/$NVIM_APPNAME/plugin` directory. Neovim will source any scripts in there automatically.

[^1]: `$XDG_CONFIG_HOME/$NVIM_APPNAME` is usually `~/.config/nvim` on Linux.

**Comment by NStefan002** [comment]
I know the questions were very basic, but the answers helped a lot. Thank you for your time.
